### ISSUE TITLE
java completion slow

### ISSUE THREAD

**Issue by notzed** [apache/netbeans#4140]
### Apache NetBeans version

Apache NetBeans 13

### What happened

Java completion is very slow with large classes due to a poor choice of algorithm selected to filter out scope visibility.

### How to reproduce

Create a class with a moderately large number of static constants or functions (>1000), import it elsewhere using a wildcard.  Try to do any completion action that might match.  It typically takes hundreds of milliseconds to show the completion window and even longer to dismiss it when you choose something (gc?).

I hit it with a vulkan binding i'm working on which puts the 3000 odd constants in a single class.

### Did this work correctly in an earlier version?

No

### Operating System

multiple systems, slackware64-current, gentoo

### JDK

openjdk 18.0.1 from openjdk.net

### Apache NetBeans packaging

Apache NetBeans provided installer

### Anything else

see: https://issues.apache.org/jira/browse/NETBEANS-6504

Sorry for now i'm not willing to fork and submit a pull request through this platform, but see the potential fixes I attached to jira if you're so inclined.

### Are you willing to submit a pull request?

No

### Code of Conduct

Yes

**Comment by mbien** [comment]
Hi @notzed, great to see you here.

I am going to take a look. I still think it would be good if you would open a PR, mostly because it would be weird for me to contribute someone else's code (even though you probably would not mind). Platforms come and go, many projects outlive them.

Are you available as reviewer at least? :)

I am going to follow your description to generate a source file as rudimentary benchmark and maybe some flame graphs later.

**Comment by mbien** [comment]
@notzed that would be the usecase, right? 3k seems to be still fairly useable while 10k causes a significant delay of several seconds, which confirms that something doesn't scale there. Worth noting is that the completion for `FieldsOfJoy10k.V*` is already slow without any star import.

```java
package test;

//import static test.FieldsOfJoy3k.*;
import static test.FieldsOfJoy10k.*;

public class LargeInterfaceCompletion {

    public static void main(String[] args) {
        System.out.println(VULKAN_);  // use completion here
        System.out.println(FieldsOfJoy10k.); // or here
    }

}
```
[FieldsOfJoy3k.java.txt](https://github.com/apache/netbeans/files/8753216/FieldsOfJoy3k.java.txt)

[FieldsOfJoy10k.java.txt](https://github.com/apache/netbeans/files/8753218/FieldsOfJoy10k.java.txt)

**Comment by notzed** [comment]
300ms isn't really what i'd call usable!  My laptop on power saving mode is much slower too.  It was enough to make me spend a couple of full days chasing this down.

The no pull request - I have political issues with the central reliance on github more than anything (google code burnt me in the past amongst other reasons), but if i decide to become more involved i'll work something out.

Reviewing your own contributed patch seems like an oxymoron to me, what's involved?

**Comment by mbien** [comment]
> 300ms isn't really what i'd call usable! My laptop on power saving mode is much slower too. It was enough to make me spend a couple of full days chasing this down.

thanks for chasing this down btw! Great work.

> The no pull request - I have political issues with the central reliance on github more than anything (google code burnt me in the past amongst other reasons),

"code forges" are temporary companions. there are projects which I moved from custom servers to sf to java.net to kenai and then to github (cvs -> svn -> git) and all of the forges did help with maintenance at some point in time. I don't think NetBeans currently overly relies on github, decisions, votes and discussion happen on the apache mailing lists. The little code details are solved in pull requests - which is fairly convenient.

If github would disappear there would be many alternatives the project could be moved to without much friction (for example gitlab, in fact someone could do it right now and host a mirror there). Also, funny enough, NetBeans uses 3 CI server clusters already (the official builds are done on jenkins / apache cluster), so the dependence there is also low.

Using a "forge" is the pragmatic choice for open source projects which are open to contributions in my opinion - doesn't matter which one (or how many in parallel).

> Reviewing your own contributed patch seems like an oxymoron to me, what's involved?

It isn't since the patch is already modified and more work will be needed most likely before we merge. I don't want to email you diffs since its fairly inefficient. Since your name is going to be on the author list I want to make sure you agree with the contribution before we merge at the very least. I see you answered the first question on the PR already - thanks :)

**Comment by istinnstudio** [comment]
I would like to make a note on this behavior, observed in large java class file with more than 15k lines and hundreds of methods/variables has degraded response on auto-completion plus general performance and secondarily also on background scanning.  I do not know if it is related only to static variables or other relations around. But larger the class, larger the chance to get performance issues.  Not sure but I think class "scaling" performance should be considered seriously in general and tested with large (>20k lines) example classes, cases of "not optimal" code that cannot or will never be separated to different classes/files will always exist.

**Comment by istinnstudio** [comment]
NB 26 is significantly faster on editing + autocompletion on very large files >20k lines. I am not sure what happened but I think this change was placed after version 24.

**Comment by mbien** [comment]
good tho hear. NB 26's java editor had some performance improvement PRs integrated https://github.com/apache/netbeans/pull/8418, https://github.com/apache/netbeans/pull/8417, https://github.com/apache/netbeans/pull/8423.

a few more are queued for NB 27 https://github.com/apache/netbeans/pulls?q=is%3Apr+author%3Ambien+label%3Aperformance+milestone%3ANB27+label%3AJava

But it will also depend on what is in those 20k lines and also which editor hints are configured to be active for that project.

`-J-Dorg.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker.level=FINE` prints some performance stats for the hints

**Comment by istinnstudio** [comment]
I can provide a part of the IDE log for 26 just for reference:

`FINE [null]: Last record repeated again.
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=4611.72ms, invocations=4729, cancelled=false}: org.netbeans.modules.java.hints.bugs.Unused
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=536.95ms, invocations=22939, cancelled=false}: org.netbeans.modules.java.hints.bugs.NPECheck
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=275.81ms, invocations=146, cancelled=false}: org.netbeans.modules.java.hints.jdk.ConvertToARM
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=202.94ms, invocations=4290, cancelled=false}: org.netbeans.modules.java.hints.threading.Tiny.canBeFinal
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=176.11ms, invocations=4729, cancelled=false}: error-in-javadoc
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=115.58ms, invocations=5, cancelled=false}: CollectionPerformance.hint
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=77.13ms, invocations=432, cancelled=false}: org.netbeans.modules.java.hints.bugs.InfiniteRecursion.run
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=66.43ms, invocations=4290, cancelled=false}: org.netbeans.modules.java.hints.HideFieldByVar
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=47.74ms, invocations=6, cancelled=false}: o.n.m.j2ee.jpa.verification.QueriesProperlyDefined
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=37.65ms, invocations=4290, cancelled=false}: org.netbeans.modules.java.hints.HideField
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=36.06ms, invocations=38325, cancelled=false}: org.netbeans.modules.java.hints.bugs.Unbalanced$Collection
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=33.95ms, invocations=132, cancelled=false}: Javac_canUseDiamond
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=31.81ms, invocations=376, cancelled=false}: UseRandomGenerator.hint
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=12.16ms, invocations=7096, cancelled=false}: org.netbeans.modules.java.hints.bugs.ThrowableNotThrown.methodInvocation
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=10.98ms, invocations=34608, cancelled=false}: org.netbeans.modules.java.hints.bugs.Unbalanced$Array
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=10.06ms, invocations=1, cancelled=false}: org.netbeans.modules.java.openjdk.jtreg.WrongSourceVersion
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=10.02ms, invocations=27046, cancelled=false}: org.netbeans.modules.java.hints.LeakingThisInConstructor
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=9.88ms, invocations=1, cancelled=false}: org.netbeans.modules.java.openjdk.jtreg.ModulesHint
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=8.26ms, invocations=9571, cancelled=false}: org.netbeans.modules.java.hints.StaticAccess
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=7.61ms, invocations=15, cancelled=false}: org.netbeans.modules.java.hints.bugs.UnusedCompoundAssignment
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=7.55ms, invocations=5, cancelled=false}: org.netbeans.modules.java.hints.jdk.ConvertToPatternInstanceOf
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=7.23ms, invocations=39, cancelled=false}: org.netbeans.modules.java.hints.PointlessBitwiseExpression
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=6.69ms, invocations=1, cancelled=false}: org.netbeans.modules.java.openjdk.jtreg.TagOrderHint
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=6.50ms, invocations=1, cancelled=false}: org.netbeans.modules.java.openjdk.jtreg.MissingRefOutputHint
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=6.18ms, invocations=24464, cancelled=false}: org.netbeans.modules.java.hints.StaticNonFinalUsedInInitialization
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=6.06ms, invocations=57, cancelled=false}: org.netbeans.modules.java.hints.suggestions.ConvertIfToSwitch
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=5.82ms, invocations=226, cancelled=false}: org.netbeans.modules.java.hints.jdk.ConvertToStringSwitch
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=5.20ms, invocations=5648, cancelled=false}: org.netbeans.modules.java.hints.bugs.UnusedAssignmentOrBranch.unusedAssignment
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=5.15ms, invocations=38, cancelled=false}: IOStreamPerformance.hint
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=5.11ms, invocations=432, cancelled=false}: org.netbeans.modules.java.hints.AddOverrideAnnotation
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=4.34ms, invocations=3, cancelled=false}: Javac_canUseLambda
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=3.99ms, invocations=4729, cancelled=false}: org.netbeans.modules.java.hints.ExportNonAccessibleElement
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=3.99ms, invocations=64, cancelled=false}: org.netbeans.modules.java.hints.jdk.IteratorToFor
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=3.85ms, invocations=3, cancelled=false}: org.netbeans.modules.java.hints.bugs.SuspiciousToArray
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=3.17ms, invocations=7100, cancelled=false}: org.netbeans.modules.java.hints.OverridableMethodCallInConstructor
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=2.85ms, invocations=20, cancelled=false}: org.netbeans.modules.java.hints.jdk.ConvertSwitchToRuleSwitch
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=2.03ms, invocations=6, cancelled=false}: org.netbeans.modules.java.hints.perf.SizeEqualsZero
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=2.00ms, invocations=15, cancelled=false}: org.netbeans.modules.java.hints.perf.ManualArrayCopy
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=1.73ms, invocations=67, cancelled=false}: org.netbeans.modules.java.hints.jdk.UseSpecificCatch
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=1.70ms, invocations=432, cancelled=false}: org.netbeans.modules.editor.htmlui.JSNI2JavaScriptBody
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=1.57ms, invocations=4290, cancelled=false}: org.netbeans.modules.java.hints.suggestions.ConstantNameHint
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=1.57ms, invocations=5, cancelled=false}: UseStandardCharsets.hint
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=1.56ms, invocations=44, cancelled=false}: org.netbeans.modules.java.hints.bugs.ArrayStringConversions
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=1.52ms, invocations=40, cancelled=false}: org.netbeans.modules.java.hints.perf.Tiny.unnecessaryTempFromString
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=1.31ms, invocations=7, cancelled=false}: org.netbeans.modules.java.hints.jdk.JoinCatches
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=1.25ms, invocations=32, cancelled=false}: org.netbeans.modules.java.hints.bugs.CollectionRemove
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=1.21ms, invocations=571, cancelled=false}: org.netbeans.modules.java.hints.bugs.UnusedAssignmentOrBranch.deadBranch
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=1.09ms, invocations=19, cancelled=false}: org.netbeans.modules.java.hints.perf.Tiny.enumMap
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=1.03ms, invocations=6, cancelled=false}: org.netbeans.modules.java.hints.jdk.AnnotationProcessors.oGSA
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=1.02ms, invocations=20, cancelled=false}: ProbableBugs.hint ... `

**Comment by istinnstudio** [comment]
sorry for opening this one again, might be related, might not, I have just found that something in code folding make performance in editor very slow, as I have used the suggested -J-Dorg.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker.level=FINE, I have seen also many entries about the code folding, after disabling in options, performance is back to normal. here is the log, more that 5k entries at once, performance drop obvious on very large editor files (so obvious that there is typing lag): 


too many logs for " Adding a fold that is identical with another previously added fold from the same FoldManager is not allowed."
If there is need for a new issue please let me know.
partial...

  ```

     New fold: [code-block] E2 <354169,358341>, desc='{...}', hash=0x334e67b8, [354169, 358341] {354170, 358340}; FoldManager: org.netbeans.modules.java.editor.fold.JavaElementFoldManager@56e02d1
Adding a fold that is identical with another previously added fold from the same FoldManager is not allowed.
Existing fold: [code-block] E2 <354169,358341>, desc='{...}', hash=0x246923b, [354169, 358341] {354170, 358340}; FoldManager: org.netbeans.modules.java.editor.fold.JavaElementFoldManager@56e02d1
     New fold: [code-block] E2 <354169,358341>, desc='{...}', hash=0x3aa2f9d6, [354169, 358341] {354170, 358340}; FoldManager: org.netbeans.modules.java.editor.fold.JavaElementFoldManager@56e02d1
Adding a fold that is identical with another previously added fold from the same FoldManager is not allowed.
Existing fold: [code-block] E2 <354169,358341>, desc='{...}', hash=0x246923b, [354169, 358341] {354170, 358340}; FoldManager: org.netbeans.modules.java.editor.fold.JavaElementFoldManager@56e02d1
     New fold: [code-block] E2 <354169,358341>, desc='{...}', hash=0x37a3e157, [354169, 358341] {354170, 358340}; FoldManager: org.netbeans.modules.java.editor.fold.JavaElementFoldManager@56e02d1
Adding a fold that is identical with another previously added fold from the same FoldManager is not allowed.
Existing fold: [code-block] E2 <354169,358341>, desc='{...}', hash=0x246923b, [354169, 358341] {354170, 358340}; FoldManager: org.netbeans.modules.java.editor.fold.JavaElementFoldManager@56e02d1
     New fold: [code-block] E2 <354169,358341>, desc='{...}', hash=0x5ef29c80, [354169, 358341] {354170, 358340}; FoldManager: org.netbeans.modules.java.editor.fold.JavaElementFoldManager@56e02d1
Adding a fold that is identical with another previously added fold from the same FoldManager is not allowed.
Existing fold: [code-block] E2 <354169,358341>, desc='{...}', hash=0x246923b, [354169, 358341] {354170, 358340}; FoldManager: org.netbeans.modules.java.editor.fold.JavaElementFoldManager@56e02d1
     New fold: [code-block] E2 <354169,358341>, desc='{...}', hash=0x2f6f8080, [354169, 358341] {354170, 358340}; FoldManager: org.netbeans.modules.java.editor.fold.JavaElementFoldManager@56e02d1

... 5K entries

   New fold: [code-block] E2 <354169,358341>, desc='{...}', hash=0x3f7377d, [354169, 358341] {354170, 358340}; FoldManager: org.netbeans.modules.java.editor.fold.JavaElementFoldManager@61e22438
Adding a fold that is identical with another previously added fold from the same FoldManager is not allowed.
Existing fold: [code-block] E2 <354169,358341>, desc='{...}', hash=0x1c367488, [354169, 358341] {354170, 358340}; FoldManager: org.netbeans.modules.java.editor.fold.JavaElementFoldManager@61e22438
     New fold: [code-block] E2 <354169,358341>, desc='{...}', hash=0x3978da70, [354169, 358341] {354170, 358340}; FoldManager: org.netbeans.modules.java.editor.fold.JavaElementFoldManager@61e22438
Adding a fold that is identical with another previously added fold from the same FoldManager is not allowed.
Existing fold: [code-block] E2 <354169,358341>, desc='{...}', hash=0x1c367488, [354169, 358341] {354170, 358340}; FoldManager: org.netbeans.modules.java.editor.fold.JavaElementFoldManager@61e22438
     New fold: [code-block] E2 <354169,358341>, desc='{...}', hash=0x1bd6f6d0, [354169, 358341] {354170, 358340}; FoldManager: org.netbeans.modules.java.editor.fold.JavaElementFoldManager@61e22438
Adding a fold that is identical with another previously added fold from the same FoldManager is not allowed.
Existing fold: [code-block] E2 <354169,358341>, desc='{...}', hash=0x1c367488, [354169, 358341] {354170, 358340}; FoldManager: org.netbeans.modules.java.editor.fold.JavaElementFoldManager@61e22438
     New fold: [code-block] E2 <354169,358341>, desc='{...}', hash=0x2577fe71, [354169, 358341] {354170, 358340}; FoldManager: org.netbeans.modules.java.editor.fold.JavaElementFoldManager@61e22438
Adding a fold that is identical with another previously added fold from the same FoldManager is not allowed.
Existing fold: [code-block] E2 <354169,358341>, desc='{...}', hash=0x1c367488, [354169, 358341] {354170, 358340}; FoldManager: org.netbeans.modules.java.editor.fold.JavaElementFoldManager@61e22438
     New fold: [code-block] E2 <354169,358341>, desc='{...}', hash=0x4a4c024a, [354169, 358341] {354170, 358340}; FoldManager: org.netbeans.modules.java.editor.fold.JavaElementFoldManager@61e22438
Adding a fold that is identical with another previously added fold from the same FoldManager is not allowed.
Existing fold: [code-block] E2 <354169,358341>, desc='{...}', hash=0x1c367488, [354169, 358341] {354170, 358340}; FoldManager: org.netbeans.modules.java.editor.fold.JavaElementFoldManager@61e22438
     New fold: [code-block] E2 <354169,358341>, desc='{...}', hash=0x568b02c8, [354169, 358341] {354170, 358340}; FoldManager: org.netbeans.modules.java.editor.fold.JavaElementFoldManager@61e22438
Adding a fold that is identical with another previously added fold from the same FoldManager is not allowed.
Existing fold: [code-block] E2 <354169,358341>, desc='{...}', hash=0x1c367488, [354169, 358341] {354170, 358340}; FoldManager: org.netbeans.modules.java.editor.fold.JavaElementFoldManager@61e22438
     New fold: [code-block] E2 <354169,358341>, desc='{...}', hash=0x7ad396d1, [354169, 358341] {354170, 358340}; FoldManager: org.netbeans.modules.java.editor.fold.JavaElementFoldManager@61e22438
Adding a fold that is identical with another previously added fold from the same FoldManager is not allowed.
Existing fold: [code-block] E2 <354169,358341>, desc='{...}', hash=0x1c367488, [354169, 358341] {354170, 358340}; FoldManager: org.netbeans.modules.java.editor.fold.JavaElementFoldManager@61e22438
     New fold: [code-block] E2 <354169,358341>, desc='{...}', hash=0x4f293429, [354169, 358341] {354170, 358340}; FoldManager: org.netbeans.modules.java.editor.fold.JavaElementFoldManager@61e22438
Adding a fold that is identical with another previously added fold from the same FoldManager is not allowed.
Existing fold: [code-block] E2 <354169,358341>, desc='{...}', hash=0x1c367488, [354169, 358341] {354170, 358340}; FoldManager: org.netbeans.modules.java.editor.fold.JavaElementFoldManager@61e22438
     New fold: [code-block] E2 <354169,358341>, desc='{...}', hash=0x4be78567, [354169, 358341] {354170, 358340}; FoldManager: org.netbeans.modules.java.editor.fold.JavaElementFoldManager@61e22438
Adding a fold that is identical with another previously added fold from the same FoldManager is not allowed.
Existing fold: [code-block] E2 <354169,358341>, desc='{...}', hash=0x1c367488, [354169, 358341] {354170, 358340}; FoldManager: org.netbeans.modules.java.editor.fold.JavaElementFoldManager@61e22438
     New fold: [code-block] E2 <354169,358341>, desc='{...}', hash=0x1fcf680f, [354169, 358341] {354170, 358340}; FoldManager: org.netbeans.modules.java.editor.fold.JavaElementFoldManager@61e22438
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=805.51ms, invocations=2253, cancelled=false}: org.netbeans.modules.java.hints.bugs.Unused
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=79.35ms, invocations=2253, cancelled=false}: error-in-javadoc
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=78.67ms, invocations=9527, cancelled=false}: org.netbeans.modules.java.hints.bugs.NPECheck
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=44.92ms, invocations=2025, cancelled=false}: org.netbeans.modules.java.hints.threading.Tiny.canBeFinal
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=32.59ms, invocations=121, cancelled=false}: Javac_canUseDiamond
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=27.82ms, invocations=224, cancelled=false}: org.netbeans.modules.java.hints.bugs.InfiniteRecursion.run
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=15.21ms, invocations=15374, cancelled=false}: org.netbeans.modules.java.hints.bugs.Unbalanced$Collection
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=15.18ms, invocations=2025, cancelled=false}: org.netbeans.modules.java.hints.HideFieldByVar
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=7.83ms, invocations=2, cancelled=false}: CollectionPerformance.hint
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=4.72ms, invocations=9551, cancelled=false}: org.netbeans.modules.java.hints.LeakingThisInConstructor
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=4.13ms, invocations=35, cancelled=false}: org.netbeans.modules.java.hints.bugs.UnusedCompoundAssignment
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=3.86ms, invocations=13414, cancelled=false}: org.netbeans.modules.java.hints.bugs.Unbalanced$Array
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=3.47ms, invocations=4334, cancelled=false}: org.netbeans.modules.java.hints.StaticAccess
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=3.30ms, invocations=2694, cancelled=false}: org.netbeans.modules.java.hints.bugs.ThrowableNotThrown.methodInvocation
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=3.22ms, invocations=1, cancelled=false}: org.netbeans.modules.java.openjdk.jtreg.ModulesHint
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=2.63ms, invocations=2160, cancelled=false}: org.netbeans.modules.java.hints.bugs.UnusedAssignmentOrBranch.unusedAssignment
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=2.61ms, invocations=1, cancelled=false}: org.netbeans.modules.java.openjdk.jtreg.MissingRefOutputHint
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=2.56ms, invocations=115, cancelled=false}: org.netbeans.modules.java.hints.jdk.ConvertToStringSwitch
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=2.54ms, invocations=1, cancelled=false}: org.netbeans.modules.java.openjdk.jtreg.TagOrderHint
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=2.51ms, invocations=1, cancelled=false}: org.netbeans.modules.java.openjdk.jtreg.WrongSourceVersion
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=2.48ms, invocations=20, cancelled=false}: org.netbeans.modules.java.hints.suggestions.ConvertIfToSwitch
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=2.33ms, invocations=2025, cancelled=false}: org.netbeans.modules.java.hints.HideField
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=1.80ms, invocations=2253, cancelled=false}: org.netbeans.modules.java.hints.ExportNonAccessibleElement
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=1.78ms, invocations=78, cancelled=false}: org.netbeans.modules.java.hints.jdk.IteratorToFor
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=1.59ms, invocations=49, cancelled=false}: org.netbeans.modules.java.hints.bugs.CollectionRemove
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=1.58ms, invocations=2694, cancelled=false}: org.netbeans.modules.java.hints.OverridableMethodCallInConstructor
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=1.50ms, invocations=9015, cancelled=false}: org.netbeans.modules.java.hints.StaticNonFinalUsedInInitialization
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=1.29ms, invocations=224, cancelled=false}: org.netbeans.modules.java.hints.AddOverrideAnnotation
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=1.28ms, invocations=224, cancelled=false}: org.netbeans.modules.java.hints.RemoveUnnecessaryReturn
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=1.21ms, invocations=2, cancelled=false}: org.netbeans.modules.java.hints.bugs.SuspiciousToArray
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.97ms, invocations=17, cancelled=false}: org.netbeans.modules.java.hints.jdk.ConvertSwitchToRuleSwitch
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.94ms, invocations=28, cancelled=false}: UseRandomGenerator.hint
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.80ms, invocations=224, cancelled=false}: org.netbeans.modules.editor.htmlui.JSNI2JavaScriptBody
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.60ms, invocations=2025, cancelled=false}: org.netbeans.modules.java.hints.suggestions.ConstantNameHint
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.58ms, invocations=9, cancelled=false}: org.netbeans.modules.java.hints.control.RedundantIf
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.53ms, invocations=381, cancelled=false}: org.netbeans.modules.java.hints.bugs.UnusedAssignmentOrBranch.deadBranch
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.30ms, invocations=570, cancelled=false}: org.netbeans.modules.java.hints.bugs.Tiny.indentation
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.27ms, invocations=2, cancelled=false}: org.netbeans.modules.java.hints.perf.Tiny.collectionsToArray
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.24ms, invocations=4, cancelled=false}: org.netbeans.modules.java.hints.perf.BoxingOfBoxingValue
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.24ms, invocations=198, cancelled=false}: org.netbeans.modules.java.hints.jdk.ConvertToTextBlock
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.23ms, invocations=2, cancelled=false}: org.netbeans.modules.java.hints.WrongPackageSuggestion
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.23ms, invocations=904, cancelled=false}: org.netbeans.modules.java.hints.InitializerCanBeStatic
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.18ms, invocations=60, cancelled=false}: org.netbeans.modules.java.hints.bugs.Tiny.varTypeDiamondOperator
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.18ms, invocations=224, cancelled=false}: org.netbeans.modules.java.hints.ClassStructure.finalPrivateMethod
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.17ms, invocations=3, cancelled=false}: o.n.m.j2ee.jpa.verification.QueriesProperlyDefined
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.16ms, invocations=4, cancelled=false}: org.netbeans.modules.java.hints.perf.Tiny.enumMap
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.14ms, invocations=381, cancelled=false}: Braces_IF
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.13ms, invocations=5, cancelled=false}: org.netbeans.modules.java.hints.jdk.UseSpecificCatch
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.13ms, invocations=54, cancelled=false}: org.netbeans.modules.java.hints.perf.Tiny.enumSet
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.12ms, invocations=5, cancelled=false}: org.netbeans.modules.java.hints.bugs.MalformedFormatString
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.12ms, invocations=224, cancelled=false}: org.netbeans.modules.java.hints.finalize.FinalizeDoesNotCallSuper
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.11ms, invocations=17, cancelled=false}: org.netbeans.modules.java.hints.jdk.ConvertToSwitchPatternInstanceOf
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.11ms, invocations=663, cancelled=false}: org.netbeans.modules.java.hints.bugs.CheckReturnValueHint
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.09ms, invocations=3, cancelled=false}: o.n.m.j2ee.jpa.verification.OnlyEntityOrMappedSuperclassCanUseIdClass
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.09ms, invocations=171, cancelled=false}: Braces_FOR_LOOP
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.07ms, invocations=1, cancelled=false}: Imports_SAME_PACKAGE
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.07ms, invocations=1, cancelled=false}: Serialization.hint
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.06ms, invocations=67, cancelled=false}: Imports_EXCLUDED
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.06ms, invocations=224, cancelled=false}: org.netbeans.modules.java.hints.finalize.FinalizeDeclared
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.06ms, invocations=3, cancelled=false}: obsoleteAnnotationSupportedSource
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.05ms, invocations=3, cancelled=false}: org.netbeans.modules.java.hints.MissingHashCode
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.05ms, invocations=262, cancelled=false}: org.netbeans.modules.java.hints.bugs.ConfusingVarargsParameter.primitiveArray
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.03ms, invocations=3, cancelled=false}: org.netbeans.modules.java.hints.jdk.AnnotationProcessors.oGSA
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.03ms, invocations=1, cancelled=false}: Imports_DEFAULT_PACKAGE
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.02ms, invocations=1, cancelled=false}: Arrays.hint
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.02ms, invocations=4, cancelled=false}: org.netbeans.modules.java.hints.MultipleLoggers
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.02ms, invocations=2, cancelled=false}: org.netbeans.modules.java.hints.threading.Tiny.nestedSynchronized
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.02ms, invocations=1, cancelled=false}: org.netbeans.modules.java.hints.IncompatibleMask
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.02ms, invocations=2, cancelled=false}: org.netbeans.modules.java.hints.EqualsHint
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.02ms, invocations=2, cancelled=false}: org.netbeans.modules.java.hints.DoubleCheck
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.01ms, invocations=2, cancelled=false}: org.netbeans.modules.java.hints.SyncOnNonFinal
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.01ms, invocations=1, cancelled=false}: org.netbeans.modules.java.hints.PointlessBitwiseExpression
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.01ms, invocations=62, cancelled=false}: unnecessaryBreakLabel
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.01ms, invocations=65, cancelled=false}: org.netbeans.modules.java.hints.threading.Tiny.volatileArray
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.01ms, invocations=4, cancelled=false}: AnnotationAsSuperInterface
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.01ms, invocations=4, cancelled=false}: org.netbeans.modules.java.hints.jdk.UnnecessaryBoxing.run
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.01ms, invocations=2, cancelled=false}: org.netbeans.modules.java.hints.control.RemoveUnnecessary.unnecessaryContinue
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.01ms, invocations=5, cancelled=false}: org.netbeans.modules.java.hints.bugs.ThrowableNotThrown.newThrowable
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.01ms, invocations=2, cancelled=false}: org.netbeans.modules.java.hints.ThisInAnonymous
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.01ms, invocations=2, cancelled=false}: EmptyStatements_FOR_LOOP
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.00ms, invocations=2, cancelled=false}: unnecessaryContinueLabel
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.00ms, invocations=1, cancelled=false}: Imports_UNUSED
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.00ms, invocations=2, cancelled=false}: EmptyStatements_BLOCK
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.00ms, invocations=2, cancelled=false}: EmptyStatements_WHILE_LOOP
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: {time=0.00ms, invocations=2, cancelled=false}: EmptyStatements_DO_WHILE_LOOP
FINE [org.netbeans.modules.java.hints.spiimpl.hints.HintsInvoker]: hint processing complete


```

**Comment by mbien** [comment]
@istinnstudio this issue here was solved, please file a new issue and submit a reproducer if you think you found a different problem. Also test with NB 27 which has release candidates available.
